File: backend/MiniMart/Models/Product.cs
namespace MiniMart.Models
{
    public class Product
    {
        public int ProductId { get; set; }

        public string ProductCode { get; set; }

        public string Barcode { get; set; }

        public string ProductName { get; set; }

        public decimal SellingPrice { get; set; }

        // Fast-read stock cache. Source of truth is the sum of active batch QuantityRemaining values.
        public int StockQuantity { get; set; }

        public int MinimumStock { get; set; }

        public byte[] RowVersion { get; set; }

        public string? Description { get; set; }

        public string? ImageUrl { get; set; }

        public bool Status { get; set; }

        public int CategoryId { get; set; }

        public Category Category { get; set; }

        public int SupplierId { get; set; }

        public Supplier Supplier { get; set; }
        public ICollection<PromotionProduct> PromotionProducts { get; set; }

        public ICollection<Batch> Batches { get; set; }
            = new List<Batch>();

        public ICollection<OrderDetail> OrderDetails { get; set; }
            = new List<OrderDetail>();

        public ICollection<InventoryTransaction> InventoryTransactions { get; set; }
            = new List<InventoryTransaction>();

        public ICollection<OrderReturnDetail> OrderReturnDetails { get; set; }
            = new List<OrderReturnDetail>();

        public ICollection<StockCountLine> StockCountLines { get; set; }
            = new List<StockCountLine>();
    }
}

File: frontend/mini_mart_management_mobile_app/lib/models/product.dart
class Product {
  const Product({
    required this.productId,
    required this.productCode,
    required this.barcode,
    required this.productName,
    required this.sellingPrice,
    required this.stockQuantity,
    required this.minimumStock,
    required this.status,
    this.description,
    this.imageUrl,
    this.categoryId,
    this.categoryName,
    this.supplierId,
    this.supplierName,
  });

  final int productId;
  final String productCode;
  final String barcode;
  final String productName;
  final double sellingPrice;
  final int stockQuantity;
  final int minimumStock;
  final bool status;
  final String? description;
  final String? imageUrl;
  final int? categoryId;
  final String? categoryName;
  final int? supplierId;
  final String? supplierName;

  factory Product.fromJson(Map<String, dynamic> j) {
    final cat = j['category'] ?? j['Category'];
    final sup = j['supplier'] ?? j['Supplier'];
    return Product(
      productId: (j['productId'] ?? j['ProductId'] ?? 0) as int,
      productCode: (j['productCode'] ?? j['ProductCode'] ?? '') as String,
      barcode: (j['barcode'] ?? j['Barcode'] ?? '') as String,
      productName: (j['productName'] ?? j['ProductName'] ?? '') as String,
      sellingPrice:
          ((j['sellingPrice'] ?? j['SellingPrice']) ?? 0).toDouble(),
      stockQuantity: (j['stockQuantity'] ?? j['StockQuantity'] ?? 0) as int,
      minimumStock: (j['minimumStock'] ?? j['MinimumStock'] ?? 0) as int,
      status: (j['status'] ?? j['Status'] ?? true) as bool,
      description: j['description'] ?? j['Description'],
      imageUrl: j['imageUrl'] ?? j['ImageUrl'],
      categoryId: cat is Map ? (cat['id'] ?? cat['Id']) as int? : null,
      categoryName:
          cat is Map ? (cat['name'] ?? cat['Name']) as String? : null,
      supplierId: sup is Map ? (sup['id'] ?? sup['Id']) as int? : null,
      supplierName:
          sup is Map ? (sup['name'] ?? sup['Name']) as String? : null,
    );
  }
}
